### ISSUE TITLE
`hs.application.get` does not find an application "exactly"

### ISSUE THREAD

**Issue by avegetablechicken** [Hammerspoon/hammerspoon#3447]
`hs.application.get("WeChat")` finds "WeChat Networking". The behavior seems not consistent with the documentation.
`hs.application.find("WeChat", true)` works well.

**Comment by Rhys-T** [comment]
Hmm - on my machine, it doesn't seem to do a partial match:
```lua
hs.application.get('Finde')
-- returns:	nil
```
Ah - that's because I'm on Catalina, so I'm stuck on an older version of Hammerspoon (0.9.97). In that version, `get` actually passes `exact=true` to `find`[^noThirdArg]:

https://github.com/Hammerspoon/hammerspoon/blob/c6855da9e09eb2612325799c9e3dc754e8296b79/extensions/application/application.lua#L102-L104 https://github.com/Hammerspoon/hammerspoon/blob/c6855da9e09eb2612325799c9e3dc754e8296b79/extensions/application/application.lua#L139

For what it's worth, I was also under the impression that `get` was supposed to be the 'exact' version - it was working that way for me, and [the documentation for `find` specifically mentions `get` as being the way to avoid partial matches][get] (and I hadn't noticed the fact that `find` takes an `exact` argument). But it appears that when #3216 was merged to fix #3206, `get` was changed to set `exact` to `false`.

I'm not actually clear on why making `get` do a non-exact match would be a fix for #3206. To me, it looks like the problem there was with trying to pass a full pathname into `hs.application.open`: `open` handles a pathname just fine ([and is documented as doing so][open]), but then when it's waiting for the app to launch, it passes that pathname along to `get`, which doesn't know what to do with pathnames whether or not `exact` is true ([nor does it claim to][get]).
```lua
hs.application.find('/Applications/Hammerspoon.app', true)
-- returns no values
hs.application.find('/Applications/Hammerspoon.app', false)
-- returns no values
hs.application.get('/Applications/Hammerspoon.app')
-- returns:	nil
```

It's entirely possible that I'm missing something here (especially since I don't have the latest version to test against), but right now it sounds to me like what should happen is:
- `get` should go back to passing `true` for `exact`.
- Either:
	- `open` should use a different method to find the app once it launches, at least for pathnames, or
	- `find` (and thus `get`) should be extended to accept a pathname as the `hint`, to match `open`.

[^noThirdArg]: And doesn't pass the third argument (`stringLiteral`) at all, because that didn't exist yet.

[find]: https://www.hammerspoon.org/docs/hs.application.html#find
[open]: https://www.hammerspoon.org/docs/hs.application.html#open
[get]: https://www.hammerspoon.org/docs/hs.application.html#get

---

@avegetablechicken: In the meantime, passing WeChat's bundle ID into either `get`, `find`, or [`applicationsForBundleID`][applicationsForBundleID] may do a better job of selecting the correct app rather than the "WeChat Networking" subprocess. From what I can find, it sounds like that ID is probably `com.tencent.xinWeChat`. If you want to make sure:
- Right-click WeChat's icon in the Finder.
- Select "Show Package Contents", then open the 'Contents' subfolder.
- Open up the 'Info.plist' file in a text editor, and look for `<key>CFBundleIdentifier</key>`.
- The bundle ID will be between `<string>` and `</string>` on the next line.

Using the bundle ID to refer to the app will also be less prone to breaking if the app name (or the OS's language setting) ever changes for any reason.

[applicationsForBundleID]: https://www.hammerspoon.org/docs/hs.application.html#applicationsForBundleID

**Comment by avegetablechicken** [comment]
> 
Thanks for your detailed reply!
